src/Hamiltonians/ho-cart-tools.jl

```diff
@@ -0,0 +1,203 @@
+    if !isnothing(target_energy) && target_energy - E0 > minimum(h.S .* h.aspect1) - 1
+        @warn "target energy higher than single particle basis size, not all blocks may be found."
+    end
+    if !isnothing(max_energy) && max_energy < E0
+        @warn "maximum requested energy lower than groundstate, not all blocks may be found."
+    end
+    if !isnothing(max_energy) && !isnothing(target_energy) && max_energy < target_energy
+        @warn "maximum requested energy lower than target energy, not all blocks may be found."
+    end
+
+    if h isa HOCartesian
+        M = h.S[1] - 1
+        if (isnothing(max_energy) && isnothing(target_energy)) ||
+            (!isnothing(max_energy) && max_energy > E0 + M) ||
+            (!isnothing(target_energy) && target_energy > E0 + M)
+            throw(ArgumentError("requested energy range not accessible by given Hamiltonian"))
+        end
+    end
+
+    if method == :vertices
+        df = get_all_blocks_vertices(h; target_energy, max_energy, max_blocks, kwargs...)
+    elseif method == :comb
+        df = get_all_blocks_comb(h; target_energy, max_energy, max_blocks, kwargs...)
+    else
+        throw(ArgumentError("invalid method."))
+    end
+
+    # consistency check
+    # if isnothing(max_blocks) && isnothing(target_energy) && isnothing(max_energy) && sum(df[!,:block_size]) ≠ dimension(h)
+    #     @warn "not all blocks were found"
+    # end
+    # return df
+end
+
+function get_all_blocks_vertices(h; 
+        target_energy = nothing, 
+        max_energy = nothing, 
+        max_blocks = nothing, 
+        save_to_file = nothing,
+        kwargs...
+    )
+    add0 = starting_address(h)
+    N = num_particles(add0)
+    P = prod(h.S)
+
+    # initialise
+    df = DataFrame()
+    block_id = 0
+    known_basis = Set{typeof(add0)}()
+    L = _binomial(P + N - 1, Val(N))
+    idx_correction = reverse(ntuple(i -> i - 1, Val(N)))
+    for i in 1:L
+        t = vertices(i, Val(N)) .- idx_correction
+        # check target energy
+        block_E0 = noninteracting_energy(h, t)
+        if !isnothing(target_energy)
+            !isapprox(block_E0, target_energy; kwargs...) && continue
+        end
+        if !isnothing(max_energy)
+            block_E0 > max_energy && continue
+        end
+        # check if known
+        addr = BoseFS(P, t .=> ones(Int, N))
+        if addr in known_basis
+            continue
+        end
+
+        # new block found
+        block_id += 1
+        t_basis = @elapsed block_basis = build_basis(h, addr)
+        for b in block_basis
+            push!(known_basis, b)
+        end
+        push!(df, (; block_id, block_E0, block_size = length(block_basis), addr, indices = t, t_basis))
+        !isnothing(save_to_file) && save_df(save_to_file, df)
+        if !isnothing(max_blocks) && block_id ≥ max_blocks
+            break
+        end
+    end
+    return df
+end
+
+# old version - issues with GC due to allocating many small vectors
```

Comment:
Does this version have any advantages? I'd document that somewhere, or remove it if not.

Reply:
It does something slightly different which may be useful sometimes, but it is not as performant so it is not the default.